src/System.Memory/src/System/SpanSortHelpers.KeysValues.TComparer.cs

```diff
@@ -0,0 +1,282 @@
+                {
+                    if (partitionSize == 1)
+                    {
+                        return;
+                    }
+                    if (partitionSize == 2)
+                    {
+                        Sort2(ref keys, ref values, lo, hi, comparer);
+                        return;
+                    }
+                    if (partitionSize == 3)
+                    {
+                        Sort3(ref keys, ref values, lo, hi - 1, hi, comparer);
+                        return;
+                    }
+                    InsertionSort(ref keys, ref values, lo, hi, comparer);
+                    return;
+                }
+
+                if (depthLimit == 0)
+                {
+                    HeapSort(ref keys, ref values, lo, hi, comparer);
+                    return;
+                }
+                depthLimit--;
+
+                // We should never reach here, unless > 3 elements due to partition size
+                int p = PickPivotAndPartition(ref keys, ref values, lo, hi, comparer);
+                // Note we've already partitioned around the pivot and do not have to move the pivot again.
+                IntroSort(ref keys, ref values, p + 1, hi, depthLimit, comparer);
+                hi = p - 1;
+            }
+        }
+
+        private static int PickPivotAndPartition<TKey, TValue, TComparer>(
+            ref TKey keys, ref TValue values, int lo, int hi,
+            TComparer comparer)
+            where TComparer : IDirectComparer<TKey>
+        {
+            Debug.Assert(comparer != null);
+            Debug.Assert(lo >= 0);
+            Debug.Assert(hi > lo);
+
+            // Compute median-of-three.  But also partition them, since we've done the comparison.
+
+            // PERF: `lo` or `hi` will never be negative inside the loop,
+            //       so computing median using uints is safe since we know 
+            //       `length <= int.MaxValue`, and indices are >= 0
+            //       and thus cannot overflow an uint. 
+            //       Saves one subtraction per loop compared to 
+            //       `int middle = lo + ((hi - lo) >> 1);`
+            int middle = (int)(((uint)hi + (uint)lo) >> 1);
+
+            // Sort lo, mid and hi appropriately, then pick mid as the pivot.
+            ref TKey keysAtMiddle = ref Sort3(ref keys, ref values, lo, middle, hi, comparer);
+
+            TKey pivot = keysAtMiddle;
+
+            int left = lo;
+            int right = hi - 1;
+            // We already partitioned lo and hi and put the pivot in hi - 1.  
+            // And we pre-increment & decrement below.
+            Swap(ref keysAtMiddle, ref Unsafe.Add(ref keys, right));
+            Swap(ref values, middle, right);
+
+            while (left < right)
+            {
+                // TODO: Would be good to be able to update local ref here
+
+                // TODO: Possible buffer over/underflow here if custom bogus comparer? What to do?
```

Comment:
This has to do bound checks. It is no ok for bogus comparer to introduce buffer overrun.

Reply:
@jkotas yes absolutely, for "outside" comparers I'm aware of that, my question is what to do? Is it ok to do the check here and throw bad comparer (via ThrowHelper) directly here? (unlike coreclr which gets an out of bounds exception and translates this to bad comparer)